libsolidity/analysis/StaticAnalyzer.cpp

```diff
@@ -48,13 +48,58 @@ void StaticAnalyzer::endVisit(ContractDefinition const&)
 
 bool StaticAnalyzer::visit(FunctionDefinition const& _function)
 {
+	m_localVarUseCount.clear();
 	m_nonPayablePublic = _function.isPublic() && !_function.isPayable();
 	return true;
 }
 
 void StaticAnalyzer::endVisit(FunctionDefinition const&)
 {
 	m_nonPayablePublic = false;
+	for (auto var = m_localVarUseCount.begin(); var != m_localVarUseCount.end(); ++var)
+	{
+		if (var->second == 0)
+		{
+			warning(var->first->location(), "Unused local variable");
+		}
+	}
+}
+
+bool StaticAnalyzer::visit(Identifier const& _identifier)
+{
+	if (auto var = dynamic_cast<VariableDeclaration const*>(_identifier.annotation().referencedDeclaration))
+	{
+		if (var->isLocalVariable() && var->name() != "")
+		{
+			if (m_localVarUseCount.find(var) != m_localVarUseCount.end())
+			{
+				m_localVarUseCount[var] += 1;
+			}
+		}
+	}
+	return true;
+}
+
+void StaticAnalyzer::endVisit(Identifier const&)
+{
+}
+
+bool StaticAnalyzer::visit(VariableDeclaration const& _variable)
+{
+	// There's no real way to get a good tracking on unnamed variables in someone's
+	// code.  These can show up in the returns clause of a function declaration,
+
+	if (_variable.isLocalVariable() && _variable.name() != "")
```

Comment:
Ah, I see. Since all the work is scoped inside functions, perhaps it would be better to set a separate flag that says "we are inside a function" and then assert that the flag is set if and only if the variable is a local variable.